src/checkers/complexity.js

```diff
@@ -0,0 +1,78 @@
+import React from "react";
+import text from "../text.json"
+
+const formatFeedback = feedback => {
+//	if (!feedback.warning && !feedback.suggestions) {
+//		return <p></p>
+//	}
+	return (
+		<>
+			<>
+				<h4>Suggestions:</h4>
+				<ul>
+					<p>{feedback}</p>
+				</ul>
+			</>
+		</>
+	);
+};
+
+const determineStrength = password => {
+	// w is password
+	// n is number of charsets in w
+	// k is the number of same-charset-substrings
+	// l is the length of password
+	// s is a bonus if w has special characters
+	// p is special character position penalty
+	// d/l is the in-dictionary penalty.
+	// s = 0.5 if w contains a special character, 0 otherwise
+	// p = 0.5 if special character is at the beginning or the end of w and w has no more 
+	// 			than 2 charsets. 0 otherwise
+	// d is the length of the substring that is a dictionary word
+	var special = "*|,\":<>[]{}`';()@&$#%";
+	var l = password.length;
+	if (l < 4) {
+		var feedback = "Password needs to be longer than 4 characters";
+		return {strength: 0, description: formatFeedback(feedback)}
+	}
+	var n = 1.0; // this is because we have the default charset = utf-8 we are gonna assume thats 1
```

Comment:
I think with their definition of charsets there's the alphanumeric charset, the numbers charset, and the special characters charset (sounds like it's not designed languages with non latin alphabets). They don't actually give a full list or a real definition, but that's my impression from the paper.

The page before the snippet in the issue has some more detail:

> As the users are aware of using different charsets to compose passwords, they are likely to put characters from the same charset together rather than mixing the them up. For example, it is more likely to have a password horse743 instead of ho7r4se3. The latter is considered more “complex” than the former and harder to crack. We find substrings in a password that are from the same charset, and count the number of such substrings. Using the same example, the number of substrings in horse743 is 2 (horse and 743), while the number of substrings in ho7r4se3 is 6 (ho, 7, r, 4, se, and 3). However, this number may be higher for a longer password than a shorter one, so we take the ratio of this number vs the password’s length as a factor to our metric.